cleanlab/experimental/datalab/issue_manager/duplicate.py

```diff
@@ -0,0 +1,147 @@
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, Any, Dict, List, Optional, Tuple, Union, cast
+
+import numpy as np
+import pandas as pd
+from sklearn.neighbors import NearestNeighbors
+from sklearn.utils.validation import check_is_fitted
+
+from cleanlab.experimental.datalab.issue_manager import IssueManager
+
+if TYPE_CHECKING:  # pragma: no cover
+    from cleanlab import Datalab
+
+
+class NearDuplicateIssueManager(IssueManager):
+    """Manages issues realted to near-duplicate examples."""
+
+    issue_name: str = "near_duplicate"
+
+    def __init__(
+        self,
+        datalab: Datalab,
+        metric: Optional[str] = "cosine",
+        threshold: Optional[float] = None,
+        k: Optional[int] = 10,
+        **_,
+    ):
+        super().__init__(datalab)
+        self.metric = metric
+        self.threshold = threshold
+        self.k = k
+        self.knn: Optional[NearestNeighbors] = None
+        self.near_duplicate_sets: List[List[int]] = []
+
+    def find_issues(
+        self,
+        features: List[str],
+        **_,
+    ) -> None:
+
+        feature_array = self._extract_embeddings(features)
+        if self.knn is None:
+            self.knn = NearestNeighbors(n_neighbors=self.k, metric=self.metric)
+
+        try:
+            check_is_fitted(self.knn)
+        except:
+            self.knn.fit(feature_array)
+
+        scores, distances = self._score_features(feature_array)
+        self.radius, self.threshold = self._compute_threshold_and_radius()
+
+        self.issues = pd.DataFrame(
+            {
+                f"is_{self.issue_name}_issue": scores < self.threshold,
+                self.issue_score_key: scores,
+            },
+        )
+
+        indices = self.knn.radius_neighbors(feature_array, self.radius, return_distance=False)
+        self.near_duplicate_sets = [
+            duplicates[duplicates != idx] for idx, duplicates in enumerate(indices)
+        ]
+        self.distances = distances
+
+        self.summary = self.get_summary(score=scores.mean())
+        self.info = self.collect_info()
+
+    def collect_info(self) -> dict:
+        issues_dict = {
+            "num_near_duplicate_issues": len(self.near_duplicate_sets),
+            "average_near_duplicate_score": self.issues[self.issue_score_key].mean(),
+            "near_duplicate_sets": self.near_duplicate_sets,
+            "radius": self.radius,
+        }
+
+        params_dict = {
+            "metric": self.metric,
+            "k": self.k,
+            "threshold": self.threshold,
+        }
+
+        knn = self.knn
+        dists, nn_ids = [array[:, 0] for array in knn.kneighbors()]  # type: ignore[union-attr]
+        weighted_knn_graph = knn.kneighbors_graph(mode="distance").toarray()  # type: ignore[union-attr]
+
+        # TODO: Reverse the order of the calls to knn.kneighbors() and knn.kneighbors_graph()
```

Comment:
Is this a quick fix? If so, maybe we could get that done before the PR is merged.

Reply:
Took some time to avoid duplicate computations as much as possible.
`dists` and `nn_ids` are now extracted from the (sparse) `weighted_knn_graph` in an efficient manner, without performing exhaustive search across all the examples for a second time.
The dense matrix is still stored in the `info` attribute.